### DISCUSSION TITLE
Directives that are Controllers--Examples?

### DISCUSSION THREAD

**Discussion by chen-ye** [lit/lit#4042] [Q&A] [answered]
The lit docs [mention](https://lit.dev/docs/composition/controllers/#controllers-and-directives) the possibility of 'controller directives'--directives that are themselves controllers. Is there an example of how this works in practice? 

I'm skimming the relevant docs, and I don't understand how a directive could implement `addController`, since the directive API doesn't expose the host. How is this supposed to work?

**Comment by augustjk** [comment, marked as answer]
I don't think we have an example in the docs site but the `@lit-labs/motion` package's `animate` directive is written as a controller directive. https://github.com/lit/lit/blob/c28cb6ed45445fb8cb5e20af5076f3d5ec9f3bea/packages/labs/motion/src/animate.ts#L227-L240

**Reply by chen-ye** [reply]
Excellent, thanks!

**Comment by bennypowers** [comment]
I'm interested in the opposite as well - controllers that surface directives. Could this be used to implement ui patterns like accessible listboxes without strong opinions about DOM structure?

**Reply by tbroyer** [reply]
Isn't `@lit-labs/tasks` such a "controller that own directive"?

**Reply by justinfagnani** [reply]
Task isn't actually. It's `.render()` method is just a plain function.

I don't know if we have an example of this in our codebase right now. We had some prototypes of observers that did this.

The basic idea is that you have a central controller and a method that returns a directive and can be used on child elements.

**Reply by augustjk** [reply]
The final form adapter POC I did here https://github.com/lit/lit/discussions/2489#discussioncomment-6105401 uses a controller that owns and surfaces a directive for binding to elements.

**Reply by bennypowers** [reply]
@nikkimk this example could be illustrative for a listbox controller.
